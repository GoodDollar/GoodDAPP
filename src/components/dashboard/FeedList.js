// @flow
import React, { useCallback, useEffect, useRef, useState } from 'react'
import { Animated, SwipeableFlatList } from 'react-native'
import * as Animatable from 'react-native-animatable'
import { get } from 'lodash'
import moment from 'moment'

import GDStore from '../../lib/undux/GDStore'
import { withStyles } from '../../lib/styles'
import { useErrorDialog } from '../../lib/undux/utils/dialog'
import userStorage from '../../lib/gundb/UserStorage'
import type { FeedEvent } from '../../lib/gundb/UserStorageClass'
import goodWallet from '../../lib/wallet/GoodWallet'
import ScrollToTopButton from '../common/buttons/ScrollToTopButton'
import logger from '../../lib/logger/pino-logger'
import { decorate, ExceptionCategory, ExceptionCode } from '../../lib/logger/exceptions'
import { CARD_OPEN, fireEvent } from '../../lib/analytics/analytics'
import FeedListItem from './FeedItems/FeedListItem'
import FeedActions from './FeedActions'
import { keyExtractor, useFeeds, VIEWABILITY_CONFIG } from './utils/feed'

const log = logger.child({ from: 'ShareButton' })

const AnimatedSwipeableFlatList = Animated.createAnimatedComponent(SwipeableFlatList)

export type FeedListProps = {
  data: any,
  onEndReached: any,
  initialNumToRender: ?number,
  store: GDStore,
  handleFeedSelection: Function,
  horizontal: boolean,
  selectedFeed: ?string,
  styles: Object,
  onScroll: Function,
}

type ItemComponentProps = {
  item: any,
  separators: {
    highlight: any,
    unhighlight: any,
  },
  index: number,
}

const getItemLayout = (_: any, index: number) => {
  const [length, separator, header] = [72, 1, 30]

  return {
    index,
    length,
    offset: (length + separator) * index + header,
  }
}

const FeedList = ({
  data,
  handleFeedSelection,
  initialNumToRender,
  onEndReached,
  onEndReachedThreshold,
  styles,
  onScroll,
  headerLarge,
  windowSize,
}: FeedListProps) => {
  const [showErrorDialog] = useErrorDialog()
  const flRef = useRef()
  const canceledFeeds = useRef([])
  const [showBounce, setShowBounce] = useState(true)
  const [displayContent, setDisplayContent] = useState(false)

  const feeds = useFeeds(data)

  const scrollToTop = () => {
    if (get(flRef, 'current._component._flatListRef.scrollToOffset')) {
      flRef.current._component._flatListRef.scrollToOffset({ offset: 0 })
    }
  }

  const pressItem = item => () => {
    if (item.type !== 'empty') {
      fireEvent(CARD_OPEN, { cardId: item.id })
      handleFeedSelection(item, true)
    }
  }

  const renderItemComponent = ({ item, separators, index }: ItemComponentProps) => (
    <FeedListItem
      key={keyExtractor(item)}
      item={item}
      separators={separators}
      fixedHeight
      onPress={pressItem(item, index + 1)}
    />
  )

  /**
   * Calls proper action depening on the feed status
   * @param {FeedEvent} item - feed item
   * @param {object} actions - wether to cancel/delete or any further action required
   */
  const handleFeedActionPress = useCallback(
    ({ id, status }: FeedEvent, actions: {}) => {
      if (actions.canCancel) {
        if (status === 'pending') {
          // if status is 'pending' trying to cancel a tx that doesn't exist will fail and may confuse the user
          log.error(
            "Current transaction is still pending, it can't be cancelled right now",
            'Pending - can`t be cancelled right now',
            new Error('Transaction is still pending'),
            {
              id,
              status,
              category: ExceptionCategory.Human,
              dialogShown: true,
            },
          )
          showErrorDialog("Current transaction is still pending, it can't be cancelled right now")
        } else if (canceledFeeds.current.includes(id)) {
          log.info('Already cancelled', id)
        } else {
          try {
            canceledFeeds.current.push(id)
            userStorage.cancelOTPLEvent(id)
            goodWallet.cancelOTLByTransactionHash(id).catch(e => {
              const uiMessage = decorate(e, ExceptionCode.E11)

              log.error('cancel payment failed - quick actions', e.message, e, {
                category: ExceptionCategory.Blockhain,
                dialogShown: true,
              })
              userStorage.updateOTPLEventStatus(id, 'pending')
              showErrorDialog(uiMessage, ExceptionCode.E11)
            })
          } catch (e) {
            const uiMessage = decorate(e, ExceptionCode.E13)

            log.error('cancel payment failed - quick actions', e.message, e, { dialogShown: true })
            canceledFeeds.current.pop()
            userStorage.updateOTPLEventStatus(id, 'pending')
            showErrorDialog(uiMessage, ExceptionCode.E13)
          }
        }
      }

      if (actions.canDelete) {
        userStorage.deleteEvent(id).catch(e => log.error('delete event failed:', e.message, e))
      }

      userStorage.userProperties.set('showQuickActionHint', false)
      setShowBounce(false)
    },
    [showErrorDialog, setShowBounce],
  )

  const renderQuickActions = useCallback(
    ({ item }) => {
      const canCancel = item && item.displayType === 'sendpending'
      const canDelete = item && item.id && item.id.indexOf('0x') === -1 && feeds.length > 1
      const hasAction = canCancel || canDelete
      const actions = { canCancel, canDelete }
      const props = { item, hasAction }

      // returning null prevents swipe action
      if (!hasAction) {
        return null
      }

      return (
        <Animatable.View animation="fadeIn" delay={750}>
          <FeedActions
            onPress={hasAction && (() => handleFeedActionPress(item, actions))}
            actionIcon={actionIcon(actions)}
            {...props}
          >
            {actionLabel(actions)}
          </FeedActions>
        </Animatable.View>
      )
    },
    [feeds],
  )

  const manageDisplayQuickActionHint = useCallback(async () => {
    // Could be string containing date to show quick action hint after - otherwise boolean
    const showQuickActionHintFlag = await userStorage.userProperties.get('showQuickActionHint')

    const _showBounce =
      typeof showQuickActionHintFlag === 'string'
        ? moment(showQuickActionHintFlag).isBefore(moment())
        : showQuickActionHintFlag

    setShowBounce(_showBounce)

    if (_showBounce) {
      await userStorage.userProperties.set(
        'showQuickActionHint',
        moment()
          .add(24, 'hours')
          .format(),
      )
    }
  }, [setShowBounce])

  useEffect(() => {
    manageDisplayQuickActionHint().finally(() => setDisplayContent(true))
  }, [])

  return displayContent ? (
    <>
      <AnimatedSwipeableFlatList
        bounceFirstRowOnMount={showBounce}
        style={styles.scrollView}
        contentContainerStyle={styles.scrollableView}
        data={feeds}
        getItemLayout={getItemLayout}
        initialNumToRender={initialNumToRender || 10}
        key="vf"
        keyExtractor={keyExtractor}
        keyboardDismissMode="on-drag"
        keyboardShouldPersistTaps="always"
        legacyImplementation={false}
        maxSwipeDistance={112}
        numColumns={1}
        onEndReached={onEndReached}
        onEndReachedThreshold={onEndReachedThreshold}
        refreshing={false}
        renderItem={renderItemComponent}
        renderQuickActions={renderQuickActions}
        viewabilityConfig={VIEWABILITY_CONFIG}
        onScroll={onScroll}
        ref={flRef}
        windowSize={windowSize}
      />
      {headerLarge ? null : <ScrollToTopButton onPress={scrollToTop} />}
    </>
  ) : null
}

const getStylesFromProps = ({ theme }) => ({
  scrollView: {
    display: 'flex',
    flexGrow: 1,
    height: 1,
  },
  scrollableView: {
    flexGrow: 1,
    display: 'flex',
    height: '100%',
  },
})

const actionLabel = ({ canDelete, canCancel }) => {
  if (canCancel) {
    return 'Cancel Payment Link'
  }

  if (canDelete) {
    return 'Delete'
  }

  return ''
}

const actionIcon = ({ canDelete, canCancel }) => {
  if (canCancel) {
    return 'close'
  }

  if (canDelete) {
    return 'trash'
  }

  return null
}

export default GDStore.withStore(withStyles(getStylesFromProps)(FeedList))
