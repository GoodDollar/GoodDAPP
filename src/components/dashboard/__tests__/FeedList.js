import React from 'react'
import renderer from 'react-test-renderer'
import FeedList from '../FeedList'
import { mockEvent } from '../__tests__/__util__'
import { withThemeProvider } from '../../../__tests__/__util__'

// Note: test renderer must be required after react-native.

describe('FeedList', () => {
  const WrappedFeedList = withThemeProvider(FeedList)
  let props = {
    updateData: () => {},
    onEndReached: () => {},
  }

  describe('Vertical rendering', () => {
    describe('With empty data', () => {
      beforeEach(() => {
        props.data = []
      })

      it('matches snapshot', async () => {
        let component

        // eslint-disable-next-line require-await
        await renderer.act(async () => (component = renderer.create(<WrappedFeedList {...props} />)))
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
      })
    })

    describe('With feed data', () => {
      beforeEach(() => {
        props.data = [
          { ...mockEvent('withdraw'), id: '0x9812619905da200c4effe8cd2ca4b2b31eeddf133f8fd283069d2e5aec3b9f8a' },
          {
            ...mockEvent('send'),
            id: '0x9812619905da200c4effe8cd2ca4b2b31eeddf133f8fd283069d2e5aec3b9f88',
          },
        ]
      })

      it('matches snapshot', async () => {
        let component

        // eslint-disable-next-line require-await
        await renderer.act(async () => (component = renderer.create(<WrappedFeedList {...props} />)))
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
      })
    })
  })
})
