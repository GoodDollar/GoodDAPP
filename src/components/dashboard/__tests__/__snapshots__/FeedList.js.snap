// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedList Vertical rendering With empty data matches snapshot 1`] = `
<div
  className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
>
  <div
    className="css-view-1dbjc4n r-WebkitOverflowScrolling-150rngu r-flexDirection-eqz5dr r-flexGrow-16y2uox r-flexShrink-1wbh5a2 r-overflowX-11yh6sk r-overflowY-1rnoaur r-transform-1sncvnh"
    onScroll={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    onWheel={[Function]}
    style={
      Object {
        "display": "flex",
        "height": "1px",
      }
    }
  >
    <div
      className="css-view-1dbjc4n r-display-6koalj r-flexGrow-16y2uox"
    >
      <div
        className="css-view-1dbjc4n"
      >
        <div
          className="css-view-1dbjc4n"
        >
          <div
            className="css-view-1dbjc4n r-borderRadius-1xfd6ze r-boxShadow-1d1oby0 r-flexDirection-18u37iz r-marginHorizontal-1ybube5 r-marginTop-knv0ih"
          >
            <div
              className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-14lw9ot r-borderRadius-1xfd6ze r-flex-13awgt0 r-justifyContent-1777fci r-overflow-1udh08x r-paddingLeft-1m04atk"
            >
              <div
                className="css-view-1dbjc4n r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-5soawk"
                style={
                  Object {
                    "backgroundColor": "rgba(238,238,238,1.00)",
                    "backgroundImage": "url(feedListItemPattern.svg)",
                  }
                }
              />
              <div
                className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1777fci r-padding-tuq35u r-width-13qz1uu"
              >
                <div
                  className="css-view-1dbjc4n r-alignSelf-173mn98 r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                  style={
                    Object {
                      "WebkitAlignItems": "flex-start",
                      "WebkitBoxAlign": "start",
                      "alignItems": "flex-start",
                      "msFlexAlign": "start",
                    }
                  }
                >
                  <div
                    className="css-view-1dbjc4n"
                    style={
                      Object {
                        "backgroundColor": "rgba(238,238,238,1.00)",
                        "borderBottomLeftRadius": "17px",
                        "borderBottomRightRadius": "17px",
                        "borderTopLeftRadius": "17px",
                        "borderTopRightRadius": "17px",
                        "height": "34px",
                        "width": "34px",
                      }
                    }
                  />
                </div>
                <div
                  className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-flexGrow-16y2uox r-justifyContent-1wtj0ep r-marginLeft-13hce6t"
                >
                  <div
                    className="css-view-1dbjc4n r-alignItems-1awozwy r-borderBottomColor-1fn3lry r-borderBottomStyle-rull8r r-borderBottomWidth-wgabs5 r-flexDirection-18u37iz r-justifyContent-1wtj0ep r-marginBottom-15zivkp r-paddingBottom-iphfwy"
                  >
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-start",
                          "WebkitBoxAlign": "start",
                          "alignItems": "flex-start",
                          "msFlexAlign": "start",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n"
                        style={
                          Object {
                            "backgroundColor": "rgba(238,238,238,1.00)",
                            "height": "10px",
                            "width": "74px",
                          }
                        }
                      />
                    </div>
                  </div>
                  <div
                    className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                  >
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-flexGrow-16y2uox r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-start",
                          "WebkitBoxAlign": "start",
                          "alignItems": "flex-start",
                          "msFlexAlign": "start",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                      >
                        <div
                          className="css-view-1dbjc4n"
                          style={
                            Object {
                              "backgroundColor": "rgba(238,238,238,1.00)",
                              "height": "18px",
                              "width": "120px",
                            }
                          }
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                      >
                        <div
                          className="css-view-1dbjc4n r-marginTop-14gqq1x"
                          style={
                            Object {
                              "backgroundColor": "rgba(238,238,238,1.00)",
                              "height": "10px",
                              "width": "74px",
                            }
                          }
                        />
                      </div>
                    </div>
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-end",
                          "WebkitBoxAlign": "end",
                          "alignItems": "flex-end",
                          "msFlexAlign": "end",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n"
                        style={
                          Object {
                            "backgroundColor": "rgba(238,238,238,1.00)",
                            "borderBottomLeftRadius": "17px",
                            "borderBottomRightRadius": "17px",
                            "borderTopLeftRadius": "17px",
                            "borderTopRightRadius": "17px",
                            "height": "34px",
                            "width": "34px",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="css-view-1dbjc4n"
        >
          <div
            className="css-view-1dbjc4n r-borderRadius-1xfd6ze r-boxShadow-1d1oby0 r-flexDirection-18u37iz r-marginHorizontal-1ybube5 r-marginTop-knv0ih"
          >
            <div
              className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-14lw9ot r-borderRadius-1xfd6ze r-flex-13awgt0 r-justifyContent-1777fci r-overflow-1udh08x r-paddingLeft-1m04atk"
            >
              <div
                className="css-view-1dbjc4n r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-5soawk"
                style={
                  Object {
                    "backgroundColor": "rgba(238,238,238,1.00)",
                    "backgroundImage": "url(feedListItemPattern.svg)",
                  }
                }
              />
              <div
                className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1777fci r-padding-tuq35u r-width-13qz1uu"
              >
                <div
                  className="css-view-1dbjc4n r-alignSelf-173mn98 r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                  style={
                    Object {
                      "WebkitAlignItems": "flex-start",
                      "WebkitBoxAlign": "start",
                      "alignItems": "flex-start",
                      "msFlexAlign": "start",
                    }
                  }
                >
                  <div
                    className="css-view-1dbjc4n"
                    style={
                      Object {
                        "backgroundColor": "rgba(238,238,238,1.00)",
                        "borderBottomLeftRadius": "17px",
                        "borderBottomRightRadius": "17px",
                        "borderTopLeftRadius": "17px",
                        "borderTopRightRadius": "17px",
                        "height": "34px",
                        "width": "34px",
                      }
                    }
                  />
                </div>
                <div
                  className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-flexGrow-16y2uox r-justifyContent-1wtj0ep r-marginLeft-13hce6t"
                >
                  <div
                    className="css-view-1dbjc4n r-alignItems-1awozwy r-borderBottomColor-1fn3lry r-borderBottomStyle-rull8r r-borderBottomWidth-wgabs5 r-flexDirection-18u37iz r-justifyContent-1wtj0ep r-marginBottom-15zivkp r-paddingBottom-iphfwy"
                  >
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-start",
                          "WebkitBoxAlign": "start",
                          "alignItems": "flex-start",
                          "msFlexAlign": "start",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n"
                        style={
                          Object {
                            "backgroundColor": "rgba(238,238,238,1.00)",
                            "height": "10px",
                            "width": "74px",
                          }
                        }
                      />
                    </div>
                  </div>
                  <div
                    className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                  >
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-flexGrow-16y2uox r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-start",
                          "WebkitBoxAlign": "start",
                          "alignItems": "flex-start",
                          "msFlexAlign": "start",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                      >
                        <div
                          className="css-view-1dbjc4n"
                          style={
                            Object {
                              "backgroundColor": "rgba(238,238,238,1.00)",
                              "height": "18px",
                              "width": "120px",
                            }
                          }
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                      >
                        <div
                          className="css-view-1dbjc4n r-marginTop-14gqq1x"
                          style={
                            Object {
                              "backgroundColor": "rgba(238,238,238,1.00)",
                              "height": "10px",
                              "width": "74px",
                            }
                          }
                        />
                      </div>
                    </div>
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-end",
                          "WebkitBoxAlign": "end",
                          "alignItems": "flex-end",
                          "msFlexAlign": "end",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n"
                        style={
                          Object {
                            "backgroundColor": "rgba(238,238,238,1.00)",
                            "borderBottomLeftRadius": "17px",
                            "borderBottomRightRadius": "17px",
                            "borderTopLeftRadius": "17px",
                            "borderTopRightRadius": "17px",
                            "height": "34px",
                            "width": "34px",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="css-view-1dbjc4n"
        >
          <div
            className="css-view-1dbjc4n r-borderRadius-1xfd6ze r-boxShadow-1d1oby0 r-flexDirection-18u37iz r-marginHorizontal-1ybube5 r-marginTop-knv0ih"
          >
            <div
              className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-14lw9ot r-borderRadius-1xfd6ze r-flex-13awgt0 r-justifyContent-1777fci r-overflow-1udh08x r-paddingLeft-1m04atk"
            >
              <div
                className="css-view-1dbjc4n r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-5soawk"
                style={
                  Object {
                    "backgroundColor": "rgba(238,238,238,1.00)",
                    "backgroundImage": "url(feedListItemPattern.svg)",
                  }
                }
              />
              <div
                className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1777fci r-padding-tuq35u r-width-13qz1uu"
              >
                <div
                  className="css-view-1dbjc4n r-alignSelf-173mn98 r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                  style={
                    Object {
                      "WebkitAlignItems": "flex-start",
                      "WebkitBoxAlign": "start",
                      "alignItems": "flex-start",
                      "msFlexAlign": "start",
                    }
                  }
                >
                  <div
                    className="css-view-1dbjc4n"
                    style={
                      Object {
                        "backgroundColor": "rgba(238,238,238,1.00)",
                        "borderBottomLeftRadius": "17px",
                        "borderBottomRightRadius": "17px",
                        "borderTopLeftRadius": "17px",
                        "borderTopRightRadius": "17px",
                        "height": "34px",
                        "width": "34px",
                      }
                    }
                  />
                </div>
                <div
                  className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-flexGrow-16y2uox r-justifyContent-1wtj0ep r-marginLeft-13hce6t"
                >
                  <div
                    className="css-view-1dbjc4n r-alignItems-1awozwy r-borderBottomColor-1fn3lry r-borderBottomStyle-rull8r r-borderBottomWidth-wgabs5 r-flexDirection-18u37iz r-justifyContent-1wtj0ep r-marginBottom-15zivkp r-paddingBottom-iphfwy"
                  >
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-start",
                          "WebkitBoxAlign": "start",
                          "alignItems": "flex-start",
                          "msFlexAlign": "start",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n"
                        style={
                          Object {
                            "backgroundColor": "rgba(238,238,238,1.00)",
                            "height": "10px",
                            "width": "74px",
                          }
                        }
                      />
                    </div>
                  </div>
                  <div
                    className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                  >
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-flexGrow-16y2uox r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-start",
                          "WebkitBoxAlign": "start",
                          "alignItems": "flex-start",
                          "msFlexAlign": "start",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                      >
                        <div
                          className="css-view-1dbjc4n"
                          style={
                            Object {
                              "backgroundColor": "rgba(238,238,238,1.00)",
                              "height": "18px",
                              "width": "120px",
                            }
                          }
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
                      >
                        <div
                          className="css-view-1dbjc4n r-marginTop-14gqq1x"
                          style={
                            Object {
                              "backgroundColor": "rgba(238,238,238,1.00)",
                              "height": "10px",
                              "width": "74px",
                            }
                          }
                        />
                      </div>
                    </div>
                    <div
                      className="css-view-1dbjc4n r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
                      style={
                        Object {
                          "WebkitAlignItems": "flex-end",
                          "WebkitBoxAlign": "end",
                          "alignItems": "flex-end",
                          "msFlexAlign": "end",
                        }
                      }
                    >
                      <div
                        className="css-view-1dbjc4n"
                        style={
                          Object {
                            "backgroundColor": "rgba(238,238,238,1.00)",
                            "borderBottomLeftRadius": "17px",
                            "borderBottomRightRadius": "17px",
                            "borderTopLeftRadius": "17px",
                            "borderTopRightRadius": "17px",
                            "height": "34px",
                            "width": "34px",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-view-1dbjc4n r-bottom-x89ehc r-height-1lh3azm r-position-u8s1d r-right-zchlnj r-width-1q42mak r-zIndex-1sg8ghl"
  >
    <div
      className="css-view-1dbjc4n"
      style={
        Object {
          "opacity": 0,
        }
      }
    >
      <div
        className="css-view-1dbjc4n r-cursor-1loqt21 r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitTransitionDuration": "0s",
            "transitionDuration": "0s",
          }
        }
        tabIndex="0"
      >
        <scrollToTop.svg />
      </div>
    </div>
  </div>
</div>
`;

exports[`FeedList Vertical rendering With feed data matches snapshot 1`] = `null`;
